pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -859,6 +871,53 @@ func isCatchAllHeaderMatch(in *networking.StringMatch) bool {
 	return catchall
 }
 
+// translateMetadataMatch translates a header match to dynamic metadata matcher. Returns nil if the header is not supported
+// or the header format is invalid for generating metadata matcher.
+//
+// The currently only supported header is x-jwt-claim for JWT claims matching. Nested claims is also supported using
+// the separator "." or "-" between claim names. For example,
```

Comment:
This is not ideal if if we can avoid it... is there any character that is a valid HTTP header but not a jwt claim? I am guessing no because claim can be any json key?

I feel like:
* Have 1 separator handles 99% of cases
* Have 2 separators handles 99.5% of cases but doubles the complexity

For example, if my claim is `"foo-bar.com"` it would not work.

Maybe an alternative is to have users URL encode any `.`s?

Reply:
> This is not ideal if if we can avoid it... is there any character that is a valid HTTP header but not a jwt claim? I am guessing no because claim can be any json key?

yeah, I don't think there would be any character that is valid in HTTP but not in jwt, it's actually the other way, you can technically put anything in the jwt claim (but practically it's mostly only ASCII characters).


> I feel like:
> 
> * Have 1 separator handles 99% of cases
> * Have 2 separators handles 99.5% of cases but doubles the complexity
> 
> For example, if my claim is `"foo-bar.com"` it would not work.
> 
> Maybe an alternative is to have users URL encode any `.`s?

I think `.` and `-` addresses the most common use cases (99.5%), Is the complexity mainly for user to understand there are 2 options for the separator? (it does not add too much complexity in implementation). I'm not sure if supporting URL encode in header is a good idea, AFAIK the header is not supposed to be url encoded.

The other alternative is probably just to support **all** valid header characters (except ALPHA and DIGIT, see the tchar definition in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6) as the separator, this is a bit similar to the multiple delimiter usage in the linux command `sed`.

@costinm @mandarjog any thoughts on this?

Reply:
updated to use `.` as separator.